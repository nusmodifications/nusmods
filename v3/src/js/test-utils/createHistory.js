// @flow

import type { ContextRouter } from 'react-router-dom';
import _ from 'lodash';

// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
import createMemoryHistory from 'history/createMemoryHistory'; // eslint-disable-line import/no-extraneous-dependencies

type MatchShape = {
  params?: { [string]: ?string },
  isExact?: boolean,
};

export default function createHistory(
  initialEntries: string | string[] = '/',
  matchParams: MatchShape = {},
): ContextRouter {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match = {
    params,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
