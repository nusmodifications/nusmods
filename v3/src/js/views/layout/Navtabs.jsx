// @flow
import React from 'react';
import { NavLink } from 'react-router-dom';
import { Calendar, Search, Settings } from 'views/components/icons/index';

import styles from './Navtabs.scss';

function Navtabs() {
  return (
    <nav className={styles.nav}>
      <NavLink className={styles.link} activeClassName={styles.linkActive} to="/timetable">
        <Calendar className={styles.icon} />
        <span className={styles.title}>Timetable</span>
      </NavLink>
      <NavLink className={styles.link} activeClassName={styles.linkActive} to="/modules">
        <Search className={styles.icon} />
        <span className={styles.title}>Browse</span>
      </NavLink>
      <NavLink className={styles.link} activeClassName={styles.linkActive} to="/settings">
        <Settings className={styles.icon} />
        <span className={styles.title}>Settings</span>
      </NavLink>
    </nav>
  );
}

export default Navtabs;
