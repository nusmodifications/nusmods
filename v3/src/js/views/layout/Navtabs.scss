@import '~styles/utils/modules-entry.scss';

/*
  For breakpoint-md (tablets), show the links as compact buttons with labels below icons.
  For breakpoint-lg (desktop), show the links as traditional link with icon to the left.
*/
.nav {
  position: fixed;
  top: $navbar-height;
  right: 0;
  left: 0;
  z-index: $nm-nav-tabs-z-index;
  // See site.scss for width on md and lg sizes
  display: flex;
  align-items: center;
  justify-content: space-around;
  height: $navtab-height;
  background-color: #fff;
  box-shadow: 0 0 15px -3px rgba(#000, 0.4);

  @include media-breakpoint-up(md) {
    flex-direction: column;
    align-items: stretch;
    justify-content: flex-start;
    margin-top: 2rem;
    box-shadow: none;
  }
}

.link {
  flex: 1 0 auto;
  margin: 0.5rem 0;
  line-height: 1;
  text-align: center;
  color: $gray-light;

  &.linkActive {
    color: theme-color(primary);
  }

  &:hover .title {
    text-decoration: underline;
  }

  @include media-breakpoint-up(sm) {
    padding: 0.5rem;
  }

  @include media-breakpoint-up(lg) {
    display: flex;
    align-items: center;
    padding: 0.8rem 0 0.8rem 2rem;
    margin: 0;
    text-align: left;
  }
}

.title {
  display: none;
  font-size: 0.7rem;

  @include media-breakpoint-up(md) {
    display: block;
  }

  @include media-breakpoint-up(lg) {
    margin-left: 0.4rem;
    font-size: 1rem;
  }
}
