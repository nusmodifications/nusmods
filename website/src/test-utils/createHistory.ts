import type { RouteComponentProps, match as Match } from 'react-router-dom';
// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import _ from 'lodash';

type MatchShape = {
  params?: { [key: string]: string | null | undefined };
  isExact?: boolean;
};

// eslint-disable-next-line @typescript-eslint/ban-types
export default function createHistory<T = {}>(
  initialEntries: string | string[] = '/',
  matchParams: MatchShape = {},
): RouteComponentProps<T> {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match: Match<T> = {
    // Not strictly type safe, but it's good enough for tests
    params: (params as unknown) as T,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
