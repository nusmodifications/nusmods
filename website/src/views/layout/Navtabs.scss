@import './nav-constants';

.nav {
  display: flex;
  justify-content: center;
  align-items: stretch;
  height: $navtab-height;
  margin: 0 auto;
}

.link {
  display: flex;
  flex-grow: 1;
  justify-content: center;
  align-items: center;
  width: 4rem;
  margin: 0 0.2rem;
  border-top: 0.2rem solid transparent;
  border-bottom: 0.2rem solid transparent;
  line-height: 1.6;
  color: var(--gray-mid);
  transition: color $desktop-entering-duration $material-deceleration-curve,
    border-bottom-color $desktop-entering-duration $material-deceleration-curve;

  &.linkActive {
    border-bottom-color: theme-color(primary);
    color: theme-color(primary);
  }

  &:hover {
    &:not(.linkActive) {
      border-bottom-color: rgba(theme-color(primary), 0.3);
    }
  }

  svg {
    flex: 0 0 auto;
    width: $navbar-icon-size;
    height: $navbar-icon-size;
    margin-bottom: 0.1rem;
  }
}

.title {
  display: none;
  margin-left: 0.4rem;
  font-size: 0.7rem;
  font-size: 1rem;
}

@include media-breakpoint-down(sm) {
  .link {
    min-width: unset;
  }
}

@include media-breakpoint-up(lg) {
  .link {
    min-width: 7em;
  }

  .title {
    display: block;
  }
}
