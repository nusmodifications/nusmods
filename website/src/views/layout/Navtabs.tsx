import React, { memo } from 'react';
import { connect } from 'react-redux';
import classnames from 'classnames';
import { BookOpen, Calendar, Clock, Heart, Map, Settings, Star, Trello } from 'react-feather';

import { Semester } from 'types/modules';
import ExternalLink from 'views/components/ExternalLink';
import { timetablePage } from 'views/routes/paths';
import { preload as preloadVenues } from 'views/venues/VenuesContainer';
import { preload as preloadContribute } from 'views/contribute/ContributeContainer';
import { PreloadingNavLink } from 'views/routes/PreloadingLink';
import { State } from 'types/state';

import styles from './Navtabs.scss';

export const NAVTAB_HEIGHT = 48;

type Props = {
  activeSemester: Semester;
  beta: boolean;
};

const tabProps = {
  className: styles.link,
  activeClassName: styles.linkActive,
};

export const NavtabsComponent = memo<Props>(({ activeSemester, beta }) => (
  <nav className={styles.nav}>
    <PreloadingNavLink {...tabProps} to="/today">
      <Clock />
      <span className={styles.title}>Today</span>
    </PreloadingNavLink>
    <PreloadingNavLink {...tabProps} to={timetablePage(activeSemester)}>
      <Calendar />
      <span className={styles.title}>Timetable</span>
    </PreloadingNavLink>
    <PreloadingNavLink
      {...tabProps}
      to={{ pathname: '/modules', search: '?sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4' }}
    >
      <BookOpen />
      <span className={styles.title}>Modules</span>
    </PreloadingNavLink>
    <PreloadingNavLink
      {...tabProps}
      to="/venues"
      onMouseOver={preloadVenues}
      onFocus={preloadVenues}
    >
      <Map />
      <span className={styles.title}>Venues</span>
    </PreloadingNavLink>
    {beta && (
      <PreloadingNavLink
        {...tabProps}
        className={classnames(tabProps.className, styles.hiddenOnMobile)}
        to="/planner"
      >
        <Trello />
        <span className={styles.title}>Planner</span>
      </PreloadingNavLink>
    )}
    <PreloadingNavLink {...tabProps} to="/settings">
      <Settings />
      <span className={styles.title}>Settings</span>
    </PreloadingNavLink>
    <PreloadingNavLink
      {...tabProps}
      className={classnames(tabProps.className, styles.hiddenOnMobile)}
      onMouseOver={preloadContribute}
      onFocus={preloadContribute}
      to="/contribute"
    >
      <Star />
      <span className={styles.title}>Contribute</span>
    </PreloadingNavLink>
    <div className={styles.divider} />
    <ExternalLink
      className={classnames(tabProps.className, styles.hiddenOnMobile)}
      href="https://nuswhispers.com"
    >
      <Heart />
      <span className={styles.title}>Whispers</span>
    </ExternalLink>
  </nav>
));

const connectedNavtabs = connect((state: State) => ({
  activeSemester: state.app.activeSemester,
  beta: !!state.settings.beta,
}))(NavtabsComponent);

export default connectedNavtabs;
