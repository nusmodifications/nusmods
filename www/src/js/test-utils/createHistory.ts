// @flow

import type { ContextRouter, LocationShape } from 'react-router-dom';
import _ from 'lodash';

// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
import createMemoryHistory from 'history/createMemoryHistory'; // eslint-disable-line import/no-extraneous-dependencies

type MatchShape = {
  params?: { [string]: ?string },
  isExact?: boolean,
};

type HistoryEntry = string | LocationShape;

export default function createHistory(
  initialEntries: HistoryEntry | $ReadOnlyArray<HistoryEntry> = '/',
  matchParams: MatchShape = {},
): ContextRouter {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match = {
    params,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
