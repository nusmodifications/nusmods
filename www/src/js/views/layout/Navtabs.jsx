// @flow
import type { State } from 'reducers';
import type { Semester } from 'types/modules';

import React from 'react';
import { connect } from 'react-redux';
import { NavLink, withRouter } from 'react-router-dom';
import classnames from 'classnames';

import { BookOpen, Calendar, Clock, Heart, Map, Settings } from 'views/components/icons';
import ExternalLink from 'views/components/ExternalLink';
import Online from 'views/components/Online';
import { timetablePage } from 'views/routes/paths';
import NavRefreshPrompt from './NavRefreshPrompt';

import styles from './Navtabs.scss';

export const NAVTAB_HEIGHT = 48;

type Props = {
  activeSemester: Semester,
  beta: boolean,
  promptRefresh: boolean,
};

export function NavtabsComponent(props: Props) {
  const tabProps = {
    className: styles.link,
    activeClassName: styles.linkActive,
    ariaCurrent: 'page',
  };

  return (
    <nav className={styles.nav}>
      {props.beta && (
        <NavLink {...tabProps} to="/today">
          <Clock />
          <span className={styles.title}>Today</span>
        </NavLink>
      )}
      <NavLink {...tabProps} to={timetablePage(props.activeSemester)}>
        <Calendar />
        <span className={styles.title}>Timetable</span>
      </NavLink>
      <NavLink {...tabProps} to="/modules">
        <BookOpen />
        <span className={styles.title}>Modules</span>
      </NavLink>
      <NavLink {...tabProps} to="/venues">
        <Map />
        <span className={styles.title}>Venues</span>
      </NavLink>
      <NavLink {...tabProps} to="/settings">
        <Settings />
        <span className={styles.title}>Settings</span>
        {props.promptRefresh && (
          <Online>
            <div
              className={classnames(styles.updateDot)}
              title="Update available"
              aria-label="Update available"
            />
          </Online>
        )}
      </NavLink>
      <div className={styles.divider} />
      <ExternalLink
        className={classnames(tabProps.className, styles.hiddenOnMobile)}
        href="https://nuswhispers.com"
      >
        <Heart />
        <span className={styles.title}>Whispers</span>
      </ExternalLink>
      {props.promptRefresh && (
        <Online>
          <NavRefreshPrompt />
        </Online>
      )}
    </nav>
  );
}

const connectedNavtabs = connect((state: State) => ({
  activeSemester: state.app.activeSemester,
  beta: state.settings.beta,
  promptRefresh: state.app.promptRefresh,
}))(NavtabsComponent);

export default withRouter(connectedNavtabs);
