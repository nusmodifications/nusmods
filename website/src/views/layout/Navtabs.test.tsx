import { shallow } from 'enzyme';

import createHistory from 'test-utils/createHistory';
import { NavtabsComponent } from './Navtabs';

describe(NavtabsComponent, () => {
  test('renders into nav element', () => {
    const navtabs = shallow(
      <NavtabsComponent activeSemester={1} beta={false} {...createHistory()} />,
    );
    expect(navtabs).toMatchSnapshot();
  });
});
