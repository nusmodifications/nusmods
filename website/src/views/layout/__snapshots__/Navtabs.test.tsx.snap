// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavtabsComponent renders into nav element 1`] = `
<nav
  className="nav"
>
  <NavLink
    activeClassName="linkActive"
    className="link"
    to="/timetable/sem-1"
  >
    <Calendar />
    <span
      className="title"
    >
      Timetable
    </span>
  </NavLink>
  <NavLink
    activeClassName="linkActive"
    className="link"
    to={
      Object {
        "pathname": "/modules",
        "search": "?sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4",
      }
    }
  >
    <BookOpen />
    <span
      className="title"
    >
      Modules
    </span>
  </NavLink>
  <NavLink
    activeClassName="linkActive"
    className="link"
    onFocus={[Function]}
    onMouseOver={[Function]}
    to="/venues"
  >
    <Map />
    <span
      className="title"
    >
      Venues
    </span>
  </NavLink>
  <NavLink
    activeClassName="linkActive"
    className="link"
    to="/settings"
  >
    <Settings />
    <span
      className="title"
    >
      Settings
    </span>
  </NavLink>
  <NavLink
    activeClassName="linkActive"
    className="link hiddenOnMobile"
    onFocus={[Function]}
    onMouseOver={[Function]}
    to="/contribute"
  >
    <Star />
    <span
      className="title"
    >
      Contribute
    </span>
  </NavLink>
  <div
    className="divider"
  />
  <ExternalLink
    className="link hiddenOnMobile"
    href="https://nuswhispers.com"
  >
    <Heart />
    <span
      className="title"
    >
      Whispers
    </span>
  </ExternalLink>
</nav>
`;
